### ISSUE TITLE
When we use isIn(emptyList) in where condition, it should not query full table.

### ISSUE THREAD

**Issue by jesusslim** [mybatis/mybatis-dynamic-sql#509]
It's been posted [here](https://github.com/mybatis/mybatis-dynamic-sql/issues/228) two years ago. But I don't think `isInRequired` or `withListEmptyCallback` is a good solution.Because if we can remember to use `isInRequired`,of course we can also remember to check empty list before `isIn`.

I can understand that you never want to build invalid sql, but in my opinion, once we use `isIn`, whatever param we pass, you should never change our sql, we don't want fixing, if you think our sql is invalid just throw exception, not change it to another wrong sql.

If we pass empty list, the sql should be `in ()`,you can't change it to query full table, **because it will be a disaster in production if we have a table with big data, there will be lots of slow queries. And if we use it in `delete`, it will delete all rows, it's a real disaster because you changed our sql wrong for "valid sql".**

And for most programmers, they don't know this issue, so every time in code review, we have to take time to check if they have checked empty list or not before `isIn`.And it's like a time bomb if someone forget to check empty list before use isIn.

I think it will be a good solution if there is a config setting to decision throw exception or query full table when isIn empty list.

**Comment by jeffgbutler** [comment]
@jesusslim see the new docs page for details on how I've addressed this issue and let me know what you think.

https://github.com/mybatis/mybatis-dynamic-sql/blob/master/src/site/markdown/docs/configuration.md

**Comment by jeffgbutler** [comment]
@Roderland Thanks for the excellent writeup!

Starting in version 1.4.1, your case [1] will throw an exception if `ageList` is empty. In this case you **do want** to query all rows in the table so you will need to configure the statement to allow the non-rendering where clause:

```java
select(c -> c.where(age, isInWhenPresent(request.getAgeList()))
   .configureStatement(cfg -> cfg.setNonRenderingWhereClauseAllowed(true)))
```

Also in version 1.4.1, your case [2] will throw an exception if `usernameList` is empty. In this case I think you **don't want** to query all rows in the table so the exception may be appropriate. But your case [3] is the best way to handle it.

Your case [3] is a good example of doing parameter validation before calling the query - that is always the best thing to do. With that case you do not need to configure the statement to allow a non-rendering where clause because you check for values in `usernameList`. Even if `ageList` is empty, the where clause will still render.

**Comment by jeffgbutler** [comment]
This change is now released in version 1.4.1. Thanks for all the feedback.
